src/Application/Services/UserService.cs

```diff
@@ -63,111 +63,150 @@ public async Task AddUserRole(User user, Role role, CancellationToken cancellati
+            };
 
-                staffMemberEntity.StaffAchievement ??= new Achievement
-                {
-                    Name = staffMemberEntity.Name,
-                    Code = AchievementHelper.GenerateCode(),
-                    Type = AchievementType.Scanned,
-                    Value = 0
-                };
-            }
+            await _dbContext.StaffMembers.AddAsync(staffMember, cancellationToken);
         }
+    }
 
+    private async Task ProcessUnclaimedAchievementsAsync(User newUser, CancellationToken cancellationToken)
+    {
         var unclaimedAchievements = await _dbContext.UnclaimedAchievements
             .TagWithContext("GetUnclaimedAchievements")
-            .Where(ua => ua.EmailAddress.ToLower() == newUser.Email.ToLower())
-            .ToListAsync();
+            .Include(ua => ua.Achievement)
+            .Where(ua => ua.EmailAddress.ToLower() == newUser.Email!.ToLower())
+            .ToListAsync(cancellationToken);
 
-        if (unclaimedAchievements.Any())
+        if (unclaimedAchievements.Count == 0) return;
+
+        var currentTime = DateTime.UtcNow;
+        
+        foreach (var unclaimedAchievement in unclaimedAchievements)
         {
-            foreach (var achievement in unclaimedAchievements)
+            newUser.UserAchievements.Add(new UserAchievement
             {
-                newUser.UserAchievements.Add(new UserAchievement
-                {
-                    Achievement = achievement.Achievement,
-                    AwardedAt = DateTime.UtcNow,
-                    CreatedUtc = DateTime.UtcNow
-                });
-
-                _dbContext.UnclaimedAchievements.Remove(achievement);
-            }
-        }
-
-        _dbContext.Users.Add(newUser);
-
-        await _dbContext.SaveChangesAsync(cancellationToken);
+                Achievement = unclaimedAchievement.Achievement,
+                AwardedAt = currentTime,
+                CreatedUtc = currentTime
+            });
 
-        _cacheService.Remove(CacheTags.NewlyUserCreated);
+            _dbContext.UnclaimedAchievements.Remove(unclaimedAchievement);
+        }
+    }
 
-        return newUser.Id;
+    private bool IsStaffEmail(string email)
+    {
+        return new MailAddress(email).Host == _staffSmtpDomain;
     }
 
     public async Task<int> GetUserId(string email, CancellationToken cancellationToken)
     {
         if (string.IsNullOrWhiteSpace(email))
-        {
-            throw new ArgumentNullException(nameof(email), "no email provided");
-        }
-
-        email = email.Trim().ToLower();
+            throw new ArgumentNullException(nameof(email), "Email cannot be null or empty");
 
-        var user = await _dbContext.Users
+        var userId = await _dbContext.Users
             .AsNoTracking()
-            .TagWithContext()
-            .Where(u => u.Email.ToLower() == email)
-            .Select(x => new { x.Id })
+            .Where(u => u.Email.ToLower() == email.Trim().ToLower())
```

Comment:
You should cache the `email` value.

```cs
email = email.Trim().ToLower();
```

```suggestion
            .Where(u => u.Email == email)
```